### PR TITLE
Adding Spark App Tasks for running in client, cluster and YARN mode

### LINKED ISSUE #9: Port sparkapp as s-c-t module
As a developer, I'd like to move *sparkapp* module from XD to s-c-t-m repo, so I can use it as task to build batch pipelines. 

**Acceptance:**
* *sparkapp* as task module (Spring Boot App) is available in *spring-cloud-task-modules* repo 
* maven coordinates include *sparkapp* with respective group/artifact/version

### PR DESCRIPTION
**Pull request by trisberg**
- adding common configuration sub-project for shared code
- adding test for local execution (cluster and YARN need to be part of integration tests)
- adding a common spark.version property

Resolves #9

### CODE REVIEW

File: sparkapp-client/pom.xml [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+				<exclusion>
+					<groupId>org.slf4j</groupId>
+					<artifactId>slf4j-log4j12</artifactId>
+				</exclusion>
+			</exclusions>
+		</dependency>
+		<dependency>
+			<groupId>com.fasterxml.jackson.core</groupId>
+			<artifactId>jackson-core</artifactId>
+			<version>2.4.4</version>
```
Comment (jvalkeal) [inline sparkapp-client/pom.xml:70]: so now the comment about these versions, should these come from parent pom?

Comment (trisberg) [inline sparkapp-client/pom.xml:70]: No, the parent pom has more recent versions that do not work, we need to override with this version

**Comment by trisberg** [comment]
# To test this with Spark 1.5.2

## Download and install:

```
wget http://mirror.nexcess.net/apache/spark/spark-1.5.2/spark-1.5.2-bin-hadoop2.6.tgz
tar xvzf spark-1.5.2-bin-hadoop2.6.tgz
cd spark-1.5.2-bin-hadoop2.6
```

Start Spark master and one worker on localhost:

```
./sbin/start-master.sh
./sbin/start-slave.sh spark://`hostname`:7077
```

You can check status at http://localhost:8080

## Copy files to HDFS:

We need the assembly jar in HDFS.

```
hadoop fs -mkdir -p /app/spark
hadoop fs -copyFromLocal lib/spark-assembly-1.5.2-hadoop2.6.0.jar /app/spark/spark-assembly-1.5.2-hadoop2.6.0.jar
```

## Run tests

```
cd <path to repo clone>
./mvnw clean install
```

### Local/client mode:

We need a local jar file for a Spark example app - there is one in a test artifact our local Maven repo - replace the {home-dir} below with the path for your home directory.

```
export SPRING_APPLICATION_JSON='{"app-name": "my-local-Pi", "app-class": "org.apache.spark.examples.JavaSparkPi", "app-jar": "sparkapp-client/target/sparkapp-client-task-1.0.0.BUILD-SNAPSHOT-tests.jar", "master": "local", "app-args": ["10"]}' 
java -jar sparkapp-client/target/sparkapp-client-task-1.0.0.BUILD-SNAPSHOT-exec.jar
```

You should see the following somewhere in all the log output:

```
Pi is roughly 3.142048
```

### Cluster mode:

We need an the app to test with available in HDFS for cluster and YARN modes:

```
hadoop fs -copyFromLocal sparkapp-client/target/sparkapp-client-task-1.0.0.BUILD-SNAPSHOT-tests.jar /app/spark/sparkapp-client-task-1.0.0.BUILD-SNAPSHOT-tests.jar
```

Now we can run the test (replace {hadoop-host} with the host for your Hadoop install and {spark-host} with you local hostname):

```
export SPRING_APPLICATION_JSON='{"app-name": "my-cluster-Pi", "app-class": "org.apache.spark.examples.JavaSparkPi", "app-jar": "hdfs://{hadoop-host}:8020/app/spark/sparkapp-client-task-1.0.0.BUILD-SNAPSHOT-tests.jar", "master": "spark://{spark-host}:7077", "deploy-mode": "cluster", "rest-url": "spark://{spark-host}:6066" , "app-args": ["10"]}'
java -jar sparkapp-cluster/target/sparkapp-cluster-task-1.0.0.BUILD-SNAPSHOT-exec.jar
```

### YARN mode:

The jar with the example app need to be in HDFS (see instructions for cluster mode above)

Now we can run the test (replace {hadoop-host} with the host for your Hadoop install):

```
SPRING_APPLICATION_JSON='{"app-name": "my-yarn-Pi", "app-class": "org.apache.spark.examples.JavaSparkPi", "app-jar": "hdfs:///app/spark/sparkapp-client-task-1.0.0.BUILD-SNAPSHOT-tests.jar", "spark-assembly-jar": "hdfs:///app/spark/spark-assembly-1.5.2-hadoop2.6.0.jar", "app-args": ["10"], "spring": {"hadoop": {"fsUri": "hdfs://{hadoop-host}:8020", "resourceManagerHost": "{hadoop-host}", "resourceManagerPort": 8032, "jobHistoryAddress": "{hadoop-host}:10020"}}}'
java -jar sparkapp-yarn/target/sparkapp-yarn-task-1.0.0.BUILD-SNAPSHOT-exec.jar
```

File: sparkapp-client/pom.xml (hunk 1 above)
Comment (jvalkeal) [inline sparkapp-client/pom.xml:70]: ok, thought it's something along those lines.

**Comment by jvalkeal** [comment]
well it's working as advertised. think I'll go into merge mode in a bit.

**Pull request merged**